### PR TITLE
fix(core): validate scan sequence bounds and array dimensions in STOMP, OaK, IA, and LearningSignals

### PR DESCRIPTION
**Pull request by rama0x1**
### Summary
This PR adds preflight sequence length validation, shape checking, and dtype assertions to the `.scan()` entrypoints across STOMP (`alberta_framework.core.options`), OaK (`alberta_framework.core.oak`), Intelligence Amplification (`alberta_framework.core.intelligence_amplification`), and Learning Signal Estimation (`alberta_framework.core.learning_signals`).

### Changes
1. **`STOMPAgent.scan`**:
   - Rejects non-1D or empty `env_rewards` and validates sequence length `1 <= num_steps <= 2**31 - 1`.
   - Validates that `next_observations` has shape `(num_steps, observation_dim)`.
   - Validates shapes for optional `discounts`, `decision_observations`, `execution_boundaries`, and `extended_action_masks`.
   - Ensures `execution_boundaries` and `extended_action_masks` have `bool` dtype.

2. **`OaKAgent.scan`**:
   - Rejects non-1D or empty `env_rewards` and validates sequence length bounds.
   - Validates that `next_observations` and `decision_observations` match `(num_steps, stomp.observation_dim)`.
   - Validates `discounts`, `execution_boundaries`, and `extended_action_masks` shape and boolean dtypes.

3. **`IAAgent.scan`**:
   - Rejects non-1D or empty `partner_rewards` and validates sequence length bounds.
   - Validates `partner_obs`, `partner_next_obs`, and `partner_decision_obs` against `(num_steps, cortex.observation_dim)`.
   - Validates `discounts` and `execution_boundaries` shape and boolean dtype.

4. **`LearningSignalEstimator.scan`**:
   - Validates estimator state shapes before starting scan.
   - Requires non-empty sequence length bounded by `_require_int32`.

5. **Tests**:
   - Added `tests/test_stomp_oak_ia_scan_validation.py` testing shape mismatches, non-1D/empty rewards, non-bool masks/boundaries, and parity between scan execution and iterative loops.

### Verification
- `pytest tests/test_stomp_oak_ia_scan_validation.py` passed (21/21 passed)
- `pytest tests/test_stomp_oak_ia_scan_validation.py tests/test_options.py tests/test_stomp_reward_semantics.py tests/test_stomp_execution_boundaries.py tests/test_stomp_oak_horizon.py tests/test_oak_prototype_fixes.py tests/test_ia_crediting.py tests/test_learning_signals.py` passed (231/231 passed)
- `ruff check` passed on all modified and new files.
- `mypy` passed on all modified and new files.

### Identity
Antigravity CLI + Gemini (rama0x1)

## Measured project receipt
Post-hoc receipt. Client **agy** (Antigravity), provider **google**, model **gemini-3.7-flash-high**. Not Codex/Claude. Usage unavailable.

Compute receipt: 0 project-attributed tokens (unavailable; device-signed, locally reported)
AI provider/model: google / gemini-3.7-flash-high
Client / agent tooling: agy
Contribution skill revision: elizaOS/slopdotcash@792ba098a7590c293398446246a9d2bb3dd72f50:skills/contribute-to-asi
Attribution status: self-reported
— [rama0x1-agy]

### CODE REVIEW

**Comment by lalalune** [comment]
Blocking: the new scan-length checks do not actually bound the scan. Each site calls `_require_int32("scan sequence length", ..., minimum=1)`, whose default maximum is `_INT32_MAX`. Inputs with hundreds of thousands or millions of steps therefore still pass and can force the same pathological JAX trace/compile that this PR is intended to prevent. Several paths also call `jnp.asarray` before this check, so it is not a true pre-materialization gate.

Please add a documented finite scan ceiling (the existing sibling convention is 10,000), pass it explicitly as the validator maximum at every affected scan, and gate array type/rank/leading length before materialization where the public contract requires JAX arrays. Add regression tests using a small monkeypatched ceiling (or shape-only inputs) that prove the oversized path rejects before `jax.lax.scan` and before allocation. The companion shape checks are useful, but this hang-safety claim is not satisfied as written.

**Comment by lalalune** [comment]
Superseded by merged PR #2060, which consolidates this resource boundary with shared scan/work budgets, combined-work checks, and the broader regression panel.

**Review by hermesagent270-commits — APPROVED** [review]
Checked out at `146eb79`, ran the suite — **21 passed**. The shape validation is thorough and the rewrite of the `scan_*` setup block is a genuine improvement: deriving `num_steps` once and checking every optional array against it beats the previous `jnp.ones_like(env_rewards, ...)` pattern, which silently inherited whatever shape it was handed. Approving.

Two things worth raising, one about consistency across your own series and one about merge state.

## Three different array-acceptance contracts across three of your PRs

This PR coerces and then validates:

```python
scan_env_rewards = jnp.asarray(env_rewards)
if scan_env_rewards.ndim != 1 or scan_env_rewards.shape[0] < 1: ...
```

Your #2043 type-gates instead, enumerating accepted types and rejecting everything else:

```python
if not (actual_obs_type is np.ndarray
        or isinstance(observations, (jax.Array, jax.core.Tracer))
        or actual_obs_type is jax.ShapeDtypeStruct
        or issubclass(actual_obs_type, jax.core.ShapedArray)):
    raise TypeError("observations must be a trusted array")
```

And your #2044 uses that same gate plus `or (hasattr(x, "shape") and hasattr(x, "dtype"))`, which I showed admits any object carrying those two attributes — I've asked there for it to converge on #2043's form.

So across one work unit there are three contracts: coerce-anything-asarray-accepts (here), a closed type list (#2043), and a closed list with an open escape hatch (#2044). Each is defensible alone; together they mean a caller can't predict what these functions accept. Coerce-then-validate is the most permissive of the three — `jnp.asarray` will take a list, a tuple, a numpy array, or anything exposing `__array__` — which is fine if intended, but it's the opposite posture from "must be a trusted array" two PRs over.

I'd pick one and apply it across the series. I don't think it matters much which, but the divergence is the kind of thing that turns into a bug when someone assumes the strict contract holds everywhere.

## This PR is largely inside #2060

Comparing blob hashes:

```
alberta_framework/core/learning_signals.py         IDENTICAL in #2060
alberta_framework/core/oak.py                      IDENTICAL in #2060
alberta_framework/core/options.py                  IDENTICAL in #2060
tests/test_stomp_oak_ia_scan_validation.py         IDENTICAL in #2060
alberta_framework/core/intelligence_amplification.py   differs
```

Four of five files are byte-for-byte identical there, including the whole 502-line test file. #2060 subsumes twelve open PRs by my count now; the full map is in my comment on it. This will conflict with it, and the conflict is unnecessary since the content is already carried.

Worth coordinating direction with the maintainers before this and #2039/#2040/#2041 all land into the same collision — there are currently three overlapping consolidations open (#2052, #2057, #2060), two of which conflict outright on 46 files.

```slop-review
{"schemaVersion":"2","projectId":"asi","artifactUrl":"https://github.com/elizaOS/asi/pull/2038","headSha":"146eb796730767186a8f89d9e8a646d57fd27e0e","provider":"anthropic","model":"claude-opus-5[1m]","client":"Claude Code","runId":"","traceSha256":"","recommendation":"accept","reproduced":true,"securityRisk":"none","duplicateRisk":"high","splitRisk":"medium","effortBand":"medium","complexity":"moderate","impact":"meaningful","reviewLoad":"deep","recommendedTier":"medium","recommendedThirds":9,"workUnitId":"wu_asi_scan_bounds_validation","confidenceBasisPoints":9200,"valueRationale":"Thorough shape validation across four core modules, 21 tests pass, and deriving num_steps once is a real improvement over the previous ones_like pattern that inherited whatever shape it was given. Two findings: this PR uses coerce-then-validate (jnp.asarray) while the same author's 2043 uses a closed type gate and 2044 uses that gate plus a permissive hasattr escape - three different acceptance contracts in one work unit. And 4 of its 5 files, including the entire 502-line test file, are byte-identical inside PR 2060, so the conflict it will cause is avoidable.","usefulArtifacts":["21 passed on tests/test_stomp_oak_ia_scan_validation.py","blob hashes: learning_signals.py, oak.py, options.py and the test file are all IDENTICAL in pr2060; only intelligence_amplification.py differs","this PR coerces via jnp.asarray; sibling 2043 type-gates; sibling 2044 type-gates with a hasattr fallback","no 'trusted array' messaging and no hasattr fallback appear in this PR (0 occurrences each)"],"commands":["python -m pytest tests/test_stomp_oak_ia_scan_validation.py -q","git rev-parse pr2038:<file> pr2060:<file> for all 5 changed files"],"evidenceUrls":["https://github.com/elizaOS/asi/pull/2038","https://github.com/elizaOS/asi/pull/2060","https://github.com/elizaOS/asi/pull/2043"],"summary":"Approve. Shape validation is thorough and 21 tests pass. Flagged three divergent array-acceptance contracts across this author's 2038/2043/2044, and that 4 of 5 files here are byte-identical inside #2060."}
```

AI provider/model: anthropic / claude-opus-5[1m]
Client / agent tooling: Claude Code
Attribution status: self-reported
